_pytest/logging.py

```diff
@@ -0,0 +1,423 @@
+        help='log format as used by the logging module.')
+    add_option_ini(
+        '--log-cli-date-format',
+        dest='log_cli_date_format', default=None,
+        help='log date format as used by the logging module.')
+    add_option_ini(
+        '--log-cli-colored',
+        dest='log_cli_colored', action='store_const',
+        const=True, default=False, type='bool',
+        help='enable colored log ouput.')
+    add_option_ini(
+        '--log-file',
+        dest='log_file', default=None,
+        help='path to a file when logging will be written to.')
+    add_option_ini(
+        '--log-file-level',
+        dest='log_file_level', default=None,
+        help='log file logging level.')
+    add_option_ini(
+        '--log-file-format',
+        dest='log_file_format', default=DEFAULT_LOG_FORMAT,
+        help='log format as used by the logging module.')
+    add_option_ini(
+        '--log-file-date-format',
+        dest='log_file_date_format', default=DEFAULT_LOG_DATE_FORMAT,
+        help='log date format as used by the logging module.')
+
+
+@contextmanager
+def logging_using_handler(handler, logger=None):
+    """Context manager that safely registers a given handler."""
+    logger = logger or logging.getLogger(logger)
+
+    if handler in logger.handlers:  # reentrancy
+        # Adding the same handler twice would confuse logging system.
+        # Just don't do that.
+        yield
+    else:
+        logger.addHandler(handler)
+        try:
+            yield
+        finally:
+            logger.removeHandler(handler)
+
+
+@contextmanager
+def catching_logs(handler, formatter=None,
+                  level=logging.NOTSET, logger=None):
+    """Context manager that prepares the whole logging machinery properly."""
+    logger = logger or logging.getLogger(logger)
+
+    if formatter is not None:
+        handler.setFormatter(formatter)
+    handler.setLevel(level)
+
+    with logging_using_handler(handler, logger):
+        orig_level = logger.level
+        logger.setLevel(min(orig_level, level))
+        try:
+            yield handler
+        finally:
+            logger.setLevel(orig_level)
+
+
+LOGLEVEL_COLORS = {
+    logging.CRITICAL: 'red',
+    logging.ERROR:  'red',
+    logging.WARNING: 'red',
+    logging.WARN:  'purple',
+    logging.INFO: 'green',
+    logging.DEBUG: 'white',
+    logging.NOTSET: 'red'
+}
+
+
+def _colorize(text, color):
+    tw = py.io.TerminalWriter()
+    tw.hasmarkup = True
+    kw = {color: True}
+    return tw.markup(text, **kw)
+
+
+class ColoredMultiLineFormatter(logging.Formatter):
+    converter = datetime.datetime.fromtimestamp
+
+    def __init__(self, *args, **kwargs):
+        self.logger_name_fmt = kwargs.pop('logger_name_fmt',
+                                          '%(name)-16s')
+        super(ColoredMultiLineFormatter, self).__init__(*args, **kwargs)
+
+        self.separator = '| '
+        self.datefmt = '%H:%M:%S.%f'
+        self._colored_levels = {
```

Comment:
set comprehensions are not supported in python 2.6 😞

Reply:
#2812 might elevate that

Reply:
BTW If you don't mind, I'll create a separate PR for the Colored Log Formatter and remove the Formatter from this commit (there are some details which need to be settled first).

Reply:
that should work out fine, thanks :)